File: exercises/3η Σειρά/Ztalloc.py
from collections import deque
import sys

class ZtallocState(object):
	def __init__(self, l, r, path):
		self.l=l
		self.r=r
		self.path=path

	def get_l(self):
		return self.l

	def get_r(self):
		return self.r

	def get_path(self):
		return self.path

	def isFinal(self, tl, tr):
		if self.l>=tl and self.r<= tr:
			return True
		else:
			return False

	def isBad(self):
		if self.r>999999:
			return True
		else:
			return False

	def next(self):
		a=[]
		a.append(ZtallocState(self.l//2, self.r//2, self.path+"h"))
		a.append(ZtallocState(self.l*3+1, self.r*3+1, self.path+"t"))
		return a

	def __eq__(self, other):
		if isinstance(other,ZtallocState):
			return self.l == other.l and self.r == other.r
		else:
			return False

	def __ne__(self, other):
		return not self.__eq__(other)

	def __str__(self):
		return str(self.l)+" "+str(self.r)+" "+self.path

	def __repr__(self):
		return str(self.l)+" "+str(self.r)+" "+self.path

	def __hash__(self):
		a = self.l+self.r
		return (a)*(a+1)//2+self.l


def ZtallocSolver(initial, l, r):
	remaining = deque([])
	seen = dict()
	seen[hash(initial)]=1
	remaining.append(initial)
	while remaining:
		a = remaining.popleft()
		if a.isFinal(l, r):
			return a
		b = a.next()
		for i in range(2):
			c = b[i]
			if not c.isBad() and not hash(c) in seen:
				remaining.append(c)
				seen[hash(c)]=1
	return None

def Ztalloc(l, r, tl, tr):
	a = ZtallocSolver(ZtallocState(l, r, ""), tl, tr)
	if a is None:
		print("IMPOSSIBLE")
	elif a.get_path()=="":
		print("EMPTY")
	else:
		print(a.get_path())

with open(sys.argv[1]) as f:
	a = [int(x) for x in next(f).split()]
	a=a[0]
	for i in range(a):
		b = [int(x) for x in next(f).split()]
		Ztalloc(b[0], b[1], b[2], b[3])
f.close()

File: exercises/3η Σειρά/Ztalloc.pl
%the queue implementation was taken from
%https://stackoverflow.com/questions/31918227/difference-lists-in-prolog-and-mutable-variables

%% empty_queue(-Queue)
% make an empty queue
empty_queue(queue(0, Q, Q)).

%% queue_head(?Queue, ?Head, ?Queue0)
% Queue, with Head removed, is Queue0
queue_head(queue(s(X), [H|Q], Q0), H, queue(X, Q, Q0)).

%% queue_last(+Queue0, +Last, -Queue)
% Queue0, with Last at its back, is Queue
queue_last(queue(X, Q, [L|Q0]), L, queue(s(X), Q, Q0)).


%the following predicate was taken from
%http://courses.softlab.ntua.gr/pl1/2019a/Exercises/read_colors_SWI.pl
read_line(Stream, L) :-
	read_line_to_codes(Stream, Line),
	atom_codes(Atom, Line),
	atomic_list_concat(Atoms, ' ', Atom),
	maplist(atom_number, Atoms, L).

read_input(Stream, 0, L, L) :-
	close(Stream), !.
read_input(Stream, N, L, NL) :-
	NN is N-1,
	read_line(Stream, A),
	append([A], L, LL),
	read_input(Stream, NN, LL, NL).

read_file(F, A) :-
	open(F, read, Stream),
	read_line(Stream, [N]),
	read_input(Stream, N, [], A).

is_final([L, R, _], TL, TR) :-
	L >= TL,
	R =< TR.

is_bad([_, R, _]) :-
	R > 999999.

next([L, R, P], N) :-
	NL1 is L//2,
	NL2 is L*3+1,
	NR1 is R//2,
	NR2 is R*3+1,
	atom_concat(P, h, NP1),
	atom_concat(P, t, NP2),
	N = [[NL1, NR1, NP1], [NL2, NR2, NP2]].

hash([L, R, _], H) :-
	A is L+R,
	H is A*(A+1)//2+L.

check1(A, T) :-
	\+ is_bad(A),
	hash(A, H),
	\+ get_assoc(H, T, _).

check(T, Q, [], T, Q).
check(T, Q, [H|L], NT, NQ) :-
	(\+ check1(H, T) -> check(T, Q, L, NT, NQ); hash(H, A), put_assoc(A, T, 1, TT), queue_last(Q, H, QQ), check(TT, QQ, L, NT, NQ)), !.

solver3(T, Q, A, TL, TR, R) :-
	next(A, B),
	check(T, Q, B, NT, NQ),
	solver2(NT, NQ, TL, TR, R), !.

solver2(_, Q, _, _, Rs) :-
	empty_queue(Q),
	Rs='IMPOSSIBLE', !.
solver2(T, Q, TL, TR, Rs) :-
	queue_head(Q, [L, R, P], QQ),
	(is_final([L, R, P], TL, TR) -> Rs=P; solver3(T, QQ, [L, R, P], TL, TR, Rs)), !.

solver1([L, R, P], TL, TR, Rs) :-
	empty_queue(Q),
	empty_assoc(T),
	queue_last(Q, [L, R, P], QQ),
	hash([L, R, P], H),
	put_assoc(H, T, 1, NT),
	(is_final([L, R, P], TL, TR) -> Rs='EMPTY'; solver2(NT, QQ, TL, TR, Rs)), !.

solve([], L, L).
solve([[L, R, TL, TR]|T], M, Rs) :-
	solver1([L, R, ''], TL, TR, A),
	append([A], M, B),
	solve(T, B, Rs).

ztalloc(F, R) :-
	read_file(F, A),
	solve(A, [], R).